qos/evm/state_archival.go

```diff
@@ -0,0 +1,190 @@
+	if a.expectedBalance == "" {
+		a.updateArchivalBalance(archivalConsensusThreshold)
+	}
+}
+
+// calculateArchivalBlockNumber determines a random archival block number based on the perceived block number.
+// The function applies the following logic:
+//   - If perceived block is below threshold, returns block 0
+//   - Otherwise, calculates a random block between minArchivalBlock and (perceivedBlockNumber - threshold)
+//   - Ensures the returned block number is never below the contract start block
+func (a *archivalState) calculateArchivalBlockNumber(perceivedBlockNumber uint64) string {
+	archivalThreshold := a.archivalCheckConfig.Threshold
+	minArchivalBlock := a.archivalCheckConfig.ContractStartBlock
+
+	var blockNumHex string
+	// Case 1: Block number is below or equal to the archival threshold
+	if perceivedBlockNumber <= archivalThreshold {
+		blockNumHex = blockNumberToHex(0)
+	} else {
+		// Case 2: Block number is above the archival threshold
+		maxBlockNumber := perceivedBlockNumber - archivalThreshold
+
+		// Ensure we don't go below the minimum archival block
+		if maxBlockNumber < minArchivalBlock {
+			blockNumHex = blockNumberToHex(minArchivalBlock)
+		} else {
+			// Generate a random block number within valid range
+			r := rand.New(rand.NewSource(time.Now().UnixNano()))
+			rangeSize := maxBlockNumber - minArchivalBlock + 1
+			blockNumHex = blockNumberToHex(minArchivalBlock + (r.Uint64() % rangeSize))
+		}
+	}
+
+	// Store the calculated block number in the archival state
+	a.blockNumberHex = blockNumHex
+	return blockNumHex
+}
+
+// updateArchivalBalance checks for consensus and updates the archival balance if it hasn't been set yet.
+// For example, if more than 5 endpoints report the same balance, the archival balance is updated to the consensus balance.
+func (a *archivalState) updateArchivalBalance(consensusThreshold int) {
+	for balance, count := range a.balanceConsensus {
+		// If we've reached the threshold, update the expected balance
+		if count >= consensusThreshold {
+			a.logger.Info().Msgf("Updating expected archival balance for block number %s to %s", a.blockNumberHex, balance)
+			a.expectedBalance = balance
+			// Reset consensus map after consensus is reached.
+			a.balanceConsensus = make(map[string]int)
+			break
+		}
+	}
+}
+
+// blockNumberToHex converts a integer block number to its hexadecimal representation.
+func blockNumberToHex(blockNumber uint64) string {
+	return fmt.Sprintf("0x%x", blockNumber)
+}
+
+// getArchivalCheckRequest returns a JSONRPC request to check the balance of:
+//   - the contract specified in `a.archivalCheckConfig.ContractAddress`
+//   - at the block number specified in `a.blockNumberHex`
+//
+// It returns false if the archival check is not enabled for the service or if the block number has not been set.
+func (a *archivalState) getArchivalCheckRequest() (jsonrpc.Request, bool) {
```

Comment:
Moving this to the checks.go could help show its role better (not ideal as it splits a structs methods, but in this specific case can improve readability)

Reply:
I've done so and moved the `archivalState` to be an argument rather than making it as method as this resolves the issue of splitting a struct method.

See commit:
https://github.com/buildwithgrove/path/pull/194/commits/cf70274f346fc5a97b0036219293e382661523cb